constraint-solver/src/solver/backtracking.rs

```diff
@@ -0,0 +1,159 @@
+            unique_assignments.len()
+        );
+
+        let mut result = BTreeMap::new();
+        for (variable, value) in unique_assignments.iter().flatten() {
+            if let Some(old_value) = result.insert(variable.clone(), *value) {
+                if old_value != *value {
+                    // Two assignments contradict each other.
+                    return Err(Error::BacktrackingError);
+                }
+            }
+        }
+
+        log::debug!("Total assignments: {}", result.len());
+        for (variable, value) in &result {
+            log::trace!("  {variable} = {value}");
+        }
+
+        Ok(result)
+    }
+
+    /// Returns all unique sets of variables that appear together in an identity
+    /// (either in an algebraic constraint or in the same field of a bus interaction),
+    /// IF the number of possible assignments is less than `MAX_BACKTRACKING_WIDTH`.
+    fn get_brute_force_candidates(&self) -> Vec<BTreeSet<V>> {
+        self.solver
+            .constraint_system
+            .iter()
+            .map(|expression| {
+                expression
+                    .referenced_variables()
+                    .cloned()
+                    .collect::<BTreeSet<_>>()
+            })
+            .unique()
+            .filter(|variables| !variables.is_empty())
+            .filter(|variables| self.has_few_possible_assignments(variables.iter().cloned()))
+            .collect()
+    }
+
+    /// Returns true if the given range constraints allow for at most `MAX_BACKTRACKING_WIDTH``
+    /// possible assignments for the given variables.
+    fn has_few_possible_assignments(&self, variables: impl Iterator<Item = V>) -> bool {
+        self.range_constraints(variables)
+            .iter()
+            .map(|rc| rc.range_width().try_into_u64())
+            .collect::<Option<Vec<_>>>()
+            .and_then(|widths| widths.iter().try_fold(1u64, |acc, &x| acc.checked_mul(x)))
+            .map(|total_width| total_width < MAX_BACKTRACKING_WIDTH)
+            .unwrap_or(false)
+    }
+
+    /// Returns the range constraints for the given variables.
+    fn range_constraints(&self, variables: impl Iterator<Item = V>) -> Vec<RangeConstraint<T>> {
+        variables
+            .map(|v| self.solver.range_constraints.get(&v))
+            .collect()
+    }
+
+    /// Goes through all possible assignments for the given variables and checks whether they satisfy
+    /// all constraints. If exactly one assignment satisfies the constraint system (and all others
+    /// lead to a contradiction), it returns that assignment.
+    /// If multiple assignments satisfy the constraint system, it returns `None`.
+    /// Returns an error if a contradiction was found.
+    fn find_unique_assignment(
+        &self,
+        variables: &BTreeSet<V>,
+    ) -> Result<Option<BTreeMap<V, T>>, Error> {
+        let mut assignment = None;
+        for assignments in self.get_all_possible_assignments(variables) {
```

Comment:
What about
```rust
match self
    .get_all_possible_assignments(variables)
    .flat_map(|assignments| self.solver.check_assignments(&assignments).then_some(assignments))
    .exactly_one()
{
  Ok(assignment) => Ok(assignment),
  Err(iter) => if iter.empty() { Err(Error::BacktrackingError) } else { Ok(None) }
}
```

Reply:
The code doesn't compile, but I implemented a similar solution. Unfortunately, `iter.empty()` does not exist, and the only way to distinguish the two error cases is to do `iter.next().is_some()` :/

We could also not care about there not being any possible assignment, but since we have the information here, I think it is good to error...